### PR TITLE
(test) Add tests for lab tests panel view

### PR DESCRIPTION
**Pull request by ODORA0**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Related Issue
https://openmrs.atlassian.net/browse/O3-4155

### CODE REVIEW

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  __esModule: true,
+  default: jest.fn(({ activePanel, groupedObservations }) => (
+    <div data-testid="panel-timeline">
+      <div>Panel: {activePanel?.name}</div>
+      <div>Observations: {Object.keys(groupedObservations).length}</div>
+    </div>
+  )),
+}));
+
+jest.mock('@carbon/react', () => {
```
Comment (CynthiaKamau) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:56]: is it possible to return the table component the same way its done in other tests?

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  createUseStore: jest.fn(() => () => ({ navGroups: [] })),
+}));
+
+jest.mock('classnames', () => ({
+  __esModule: true,
+  default: jest.fn((...args) => args.join(' ')),
+}));
+
+jest.mock('@openmrs/esm-patient-common-lib', () => ({
+  EmptyState: jest.fn(({ headerTitle }) => <div data-testid="empty-state">{headerTitle}</div>),
```
Comment (CynthiaKamau) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:30]: dtto, you can mock an empty dataset and return the `<EmptyState>` container just like we do it in other tests

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:1]: Could we rename this to `panel-view.test.tsx` per our [naming conventions](https://o3-docs.openmrs.org/docs/coding-conventions#naming)?

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx [hunk 3]
```diff
@@ -0,0 +1,216 @@
+import React from 'react';
+import { render, screen, fireEvent, within } from '@testing-library/react';
+import { useTranslation } from 'react-i18next';
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:3]: We shouldn't need this import.

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+import PanelView from './panel-view.component';
+import usePanelData from './usePanelData';
+import classNames from 'classnames';
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:6]: Ditto.

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    unsubscribe: jest.fn(),
+    setState: jest.fn(),
+  })),
+  createUseStore: jest.fn(() => () => ({ navGroups: [] })),
+}));
+
+jest.mock('classnames', () => ({
+  __esModule: true,
+  default: jest.fn((...args) => args.join(' ')),
+}));
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:27]: We shouldn't need either of these mocks. For the framework, we can rely on the test stubs from the framework mock. See https://o3-docs.openmrs.org/docs/frontend-modules/unit-and-integration-testing#mocking-patterns.

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  )),
+}));
+
+jest.mock('./usePanelData');
+
+jest.mock('react-i18next', () => ({
+  useTranslation: jest.fn().mockReturnValue({
+    t: (key) => key,
+  }),
+}));
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:44]: Shouldn't need this mock either. `react-i18next` gets mocked in the jest config's moduleNameMapper to a file in the top level `__mocks__` directory.

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx (hunk 1 above)
Comment (denniskigen) [reply packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:56]: Yeah, we shouldn't mock `carbon/react` either.

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        value: '35',
+        effectiveDateTime: '2024-01-02T10:00:00Z',
+      },
+    ],
+  },
+};
+
+describe('PanelView', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:171]: This gets taken care of [automatically](https://github.com/openmrs/openmrs-esm-patient-chart/blob/main/jest.config.js#L6).

**Review by denniskigen — APPROVED** [review]
Thanks for taking a stab at this, @ODORA0. I've left some comments.

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.tsx [hunk 8]
```diff
@@ -90,7 +90,7 @@ const PanelView: React.FC<PanelViewProps> = ({ expanded, testUuid, basePath, typ
               <EmptyState displayText={t('panels', 'panels')} headerTitle={t('noPanelsFound', 'No panels found')} />
             )
           ) : (
-            <DataTableSkeleton columns={3} />
+            <DataTableSkeleton data-testid="data-table-skeleton" columns={3} />
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.tsx:93]: Prefer using the `progressbar` ARIA role over a testid:

```tsx
<DataTableSkeleton  columns={3} role="progressbar" />
```

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.tsx [hunk 9]
```diff
@@ -137,7 +137,7 @@ const PanelView: React.FC<PanelViewProps> = ({ expanded, testUuid, basePath, typ
                 <EmptyState displayText={t('panels', 'panels')} headerTitle={t('noPanelsFound', 'No panels found')} />
               )
             ) : (
-              <DataTableSkeleton columns={3} />
+              <DataTableSkeleton data-testid="data-table-skeleton" columns={3} />
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.tsx:140]:
```suggestion
              <DataTableSkeleton  columns={3} role="progressbar" />
```

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.tsx [hunk 10]
```diff
@@ -147,7 +147,7 @@ const PanelView: React.FC<PanelViewProps> = ({ expanded, testUuid, basePath, typ
       >
         <div className={styles.stickySection}>
           {isLoading ? (
-            <DataTableSkeleton columns={3} />
+            <DataTableSkeleton data-testid="data-table-skeleton" columns={3} />
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.tsx:150]: Ditto

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    (usePanelData as jest.Mock).mockReturnValue({
+      panels: [],
+      isLoading: false,
+      groupedObservations: {},
+    });
+
+    render(<PanelView expanded={false} testUuid="" type="" basePath="/test" patientUuid="test-patient" />);
+
+    expect(screen.getByTestId('empty-state')).toBeInTheDocument();
+    expect(screen.getByText('noPanelsFound')).toBeInTheDocument();
```
Comment (denniskigen) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.component.test.tsx:196]:
```suggestion
    expect(screen.getByText(/no panels found/i)).toBeInTheDocument();
    expect(screen.getByText(/There are no panels to display for this patient/i)).toBeInTheDocument();
```

File: packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.test.tsx [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    <div>
+      <h1>{headerTitle}</h1>
+      <p>There are no {displayText} to display for this patient</p>
+    </div>
+  ),
+}));
+
+jest.mock('../panel-timeline/panel-timeline-component', () => ({
+  __esModule: true,
+  default: ({ activePanel }) => <div data-testid="panel-timeline-component">{activePanel?.name}</div>,
```
Comment (CynthiaKamau) [inline packages/esm-patient-tests-app/src/test-results/panel-view/panel-view.test.tsx:39]: why did you add this ? it is used somewhere ?

**Review by denniskigen — APPROVED** [review]
Sorry, I didn't main to rain on your parade here, @ODORA0. I meant to suggest some changes but then I went down a rabbit hole and wound up changing too many things. Hopefully it's all clear and fine with you!
